File: python/day9.py
DIRECTIONS = [(-1, 0), (1, 0), (0, -1), (0, 1)]

def adjacent_locations(x, y, width, height):
    for (dx, dy) in DIRECTIONS:
        (nx, ny) = (x + dx, y + dy)

        if nx >= 0 and ny >= 0 and nx < width and ny < height:
            yield (nx, ny)

def solve():
    with open("problems/problem9") as f:
        lines = f.readlines()

    grid = [[int(x) for x in line.strip()] for line in lines if not len(line.strip()) == 0]
    width = len(grid[0])
    height = len(grid)

    prob1 = 0
    basin_sizes = []

    for x in range(width):
        for y in range(height):
            lowpoint = True
            val = grid[y][x]

            for nx, ny in adjacent_locations(x, y, width, height):
                if grid[ny][nx] <= val:
                    lowpoint = False
                    break

            if not lowpoint:
                continue

            prob1 += val + 1

            basin = set()
            to_be_checked = [(x, y)]
            while len(to_be_checked) > 0:
                cx, cy = to_be_checked.pop()
                basin.add((cx, cy))

                val = grid[cy][cx]
                for nx, ny in adjacent_locations(cx, cy, width, height):
                    nval = grid[ny][nx]

                    if nval < 9 and nval > val and (nx, ny) not in basin:
                        to_be_checked.append((nx, ny))

            basin_sizes.append(len(basin))

    
    print("Problem 1:", prob1)
    
    basin_sizes.sort(reverse=True)
    prob2 = basin_sizes[0] * basin_sizes[1] * basin_sizes[2]

    print("Problem 2:", prob2)

if __name__ == '__main__':
    solve()

File: src/day9.rs
use std::collections::HashSet;

use crate::get_input_lines;

const INPUT: &str = "problems/problem9";

pub fn adjacent_locations(
    x: usize,
    y: usize,
    width: usize,
    height: usize,
) -> impl Iterator<Item = (usize, usize)> {
    [(-1, 0), (1, 0), (0, -1), (0, 1)]
        .iter()
        .filter_map(move |&(dx, dy)| {
            let (nx, ny) = ((x as isize) + dx, (y as isize) + dy);

            if nx >= 0 && ny >= 0 && nx < width as isize && ny < height as isize {
                Some((nx as usize, ny as usize))
            } else {
                None
            }
        })
}

pub fn solve() -> crate::Result<()> {
    let lines = get_input_lines(INPUT)?;

    let mut grid: Vec<u8> = Vec::new();
    let mut width = 0;

    for line in lines {
        let parsed = line
            .trim()
            .chars()
            .map(|c| char::to_digit(c, 10).unwrap() as u8);

        grid.extend(parsed);

        if width == 0 {
            width = grid.len();
        }
    }

    let height = grid.len() / width;

    let mut prob1: u64 = 0;

    let mut basin_sizes: Vec<usize> = Default::default();

    for x in 0..width {
        for y in 0..height {
            let mut lowpoint = true;
            let val = grid[x + y * width];

            for (nx, ny) in adjacent_locations(x, y, width, height) {
                if grid[nx as usize + ny as usize * width] <= val {
                    lowpoint = false;
                    break;
                }
            }

            if !lowpoint {
                continue;
            }

            prob1 += val as u64 + 1;

            let mut basin: HashSet<(usize, usize)> = Default::default();
            let mut to_be_checked: Vec<(usize, usize)> = vec![(x, y)];

            while let Some((x, y)) = to_be_checked.pop() {
                basin.insert((x, y));

                let val = grid[x + y * width];

                for (nx, ny) in adjacent_locations(x, y, width, height) {
                    let nval = grid[nx + ny * width];

                    if nval < 9 && nval > val && !basin.contains(&(nx, ny)) {
                        to_be_checked.push((nx, ny));
                    }
                }
            }

            basin_sizes.push(basin.len());
        }
    }

    println!("Problem 1: {}", prob1);

    basin_sizes.sort();
    let prob2: u64 = basin_sizes
        .iter()
        .rev()
        .take(3)
        .map(|x| *x as u64)
        .product();

    println!("Problem 2: {}", prob2);

    Ok(())
}
